File: dev/tools/test/analyze_imports_test.dart
// Copyright 2014 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import 'dart:io';

import '../analyze_imports.dart';

void main() {
  final Directory tempDir = Directory.systemTemp.createTempSync('analyze_imports_test_');
  final File tempFile = File('${tempDir.path}/test_imports.csv');

  try {
    tempFile.writeAsStringSync(
      'source_file,imported_file,symbol\n'
      'A.dart,B.dart,sym1\n'
      'A.dart,B.dart,sym2\n'
      'B.dart,C.dart,sym3\n'
      'C.dart,A.dart,sym4\n'
      'C.dart,D.dart,sym5\n'
      'D.dart,E.dart,sym6\n'
      'F.dart,G.dart,sym7\n'
      'G.dart,F.dart,sym8\n'
      'H.dart,D.dart,sym9\n',
    );

    final ImportAnalyzer analyzer = ImportAnalyzer(tempFile.path);
    analyzer.load();

    // 1. Test direct imports
    final List<DirectImportRecord> direct = analyzer.computeDirectImportSymbolCounts();
    if (direct.length != 8) {
      throw StateError('Expected 8 direct import pairs, got ${direct.length}');
    }
    final DirectImportRecord ab = direct.firstWhere((r) => r.sourceFile == 'A.dart' && r.importedFile == 'B.dart');
    if (ab.symbolCount != 2 || ab.symbols[0] != 'sym1' || ab.symbols[1] != 'sym2') {
      throw StateError('A.dart -> B.dart symbol mismatch: ${ab.symbols}');
    }

    // 2. Test transitive imports
    final Map<String, List<String>> transitive = analyzer.computeTransitiveImports();
    final List<String>? transA = transitive['A.dart'];
    if (transA == null || transA.join(',') != 'A.dart,B.dart,C.dart,D.dart,E.dart') {
      throw StateError('Transitive imports for A.dart mismatch: $transA');
    }
    final List<String>? transH = transitive['H.dart'];
    if (transH == null || transH.join(',') != 'D.dart,E.dart') {
      throw StateError('Transitive imports for H.dart mismatch: $transH');
    }

    // 3. Test cycles
    final List<List<String>> cycles = analyzer.computeImportCycles();
    if (cycles.length != 2) {
      throw StateError('Expected 2 cycles, got ${cycles.length}');
    }
    if (cycles[0].join(',') != 'A.dart,B.dart,C.dart') {
      throw StateError('Cycle 1 mismatch: ${cycles[0]}');
    }
    if (cycles[1].join(',') != 'F.dart,G.dart') {
      throw StateError('Cycle 2 mismatch: ${cycles[1]}');
    }

    // Verify cycle property: every file in cycle reaches all files in cycle
    for (final List<String> cycle in cycles) {
      final Set<String> cycleSet = cycle.toSet();
      for (final String f in cycle) {
        final Set<String> trans = transitive[f]!.toSet();
        if (!cycleSet.every(trans.contains)) {
          throw StateError('File $f in cycle did not transitively reach all cycle nodes: $cycle');
        }
      }
    }

    stdout.writeln('All analyze_imports Dart tests passed successfully!');
  } finally {
    tempDir.deleteSync(recursive: true);
  }
}

File: dev/tools/test/analyze_imports_test.py
#!/usr/bin/env python3
"""
Unit tests for analyze_imports.py.
"""

import os
import sys
import tempfile
import unittest

# Import analyze_imports module
sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), "..")))
import analyze_imports


class TestAnalyzeImports(unittest.TestCase):
    def setUp(self):
        self.temp_file = tempfile.NamedTemporaryFile("w+", delete=False, suffix=".csv")
        # Write test CSV graph:
        # A -> B (sym1, sym2)
        # B -> C (sym3)
        # C -> A (sym4)  [A, B, C form a 3-node cycle]
        # C -> D (sym5)
        # D -> E (sym6)
        # F -> G (sym7)
        # G -> F (sym8)  [F, G form a 2-node cycle]
        # H -> D (sym9)  [H is a linear leaf importer]
        self.temp_file.write(
            "source_file,imported_file,symbol\n"
            "A.dart,B.dart,sym1\n"
            "A.dart,B.dart,sym2\n"
            "B.dart,C.dart,sym3\n"
            "C.dart,A.dart,sym4\n"
            "C.dart,D.dart,sym5\n"
            "D.dart,E.dart,sym6\n"
            "F.dart,G.dart,sym7\n"
            "G.dart,F.dart,sym8\n"
            "H.dart,D.dart,sym9\n"
        )
        self.temp_file.close()

    def tearDown(self):
        if os.path.exists(self.temp_file.name):
            os.remove(self.temp_file.name)

    def test_direct_imports_and_symbols(self):
        analyzer = analyze_imports.ImportAnalyzer(self.temp_file.name)
        direct = analyzer.compute_direct_import_symbol_counts()
        direct_map = {(src, imp): (cnt, syms) for src, imp, cnt, syms in direct}

        self.assertEqual(direct_map[("A.dart", "B.dart")], (2, ["sym1", "sym2"]))
        self.assertEqual(direct_map[("B.dart", "C.dart")], (1, ["sym3"]))
        self.assertEqual(direct_map[("C.dart", "A.dart")], (1, ["sym4"]))
        self.assertEqual(direct_map[("C.dart", "D.dart")], (1, ["sym5"]))
        self.assertEqual(direct_map[("D.dart", "E.dart")], (1, ["sym6"]))
        self.assertEqual(direct_map[("F.dart", "G.dart")], (1, ["sym7"]))
        self.assertEqual(direct_map[("G.dart", "F.dart")], (1, ["sym8"]))
        self.assertEqual(direct_map[("H.dart", "D.dart")], (1, ["sym9"]))

    def test_transitive_imports(self):
        analyzer = analyze_imports.ImportAnalyzer(self.temp_file.name)
        transitive = analyzer.compute_transitive_imports()

        # A reaches B, C, A, D, E
        self.assertEqual(transitive["A.dart"], ["A.dart", "B.dart", "C.dart", "D.dart", "E.dart"])
        # B reaches C, A, B, D, E
        self.assertEqual(transitive["B.dart"], ["A.dart", "B.dart", "C.dart", "D.dart", "E.dart"])
        # C reaches A, B, C, D, E
        self.assertEqual(transitive["C.dart"], ["A.dart", "B.dart", "C.dart", "D.dart", "E.dart"])
        # D reaches E
        self.assertEqual(transitive["D.dart"], ["E.dart"])
        # F reaches F, G
        self.assertEqual(transitive["F.dart"], ["F.dart", "G.dart"])
        # G reaches F, G
        self.assertEqual(transitive["G.dart"], ["F.dart", "G.dart"])
        # H reaches D, E
        self.assertEqual(transitive["H.dart"], ["D.dart", "E.dart"])

    def test_import_cycles(self):
        analyzer = analyze_imports.ImportAnalyzer(self.temp_file.name)
        cycles = analyzer.compute_import_cycles()

        # Expect 2 cycles: [A, B, C] of size 3 and [F, G] of size 2
        self.assertEqual(len(cycles), 2)
        self.assertEqual(cycles[0], ["A.dart", "B.dart", "C.dart"])
        self.assertEqual(cycles[1], ["F.dart", "G.dart"])

        # Check property: importing any single file in cycle reaches all others
        for cycle in cycles:
            cycle_set = set(cycle)
            for f in cycle:
                trans = set(analyzer.compute_transitive_imports()[f])
                self.assertTrue(cycle_set.issubset(trans))


if __name__ == "__main__":
    unittest.main()
